File: example/cpp/xpt/udp_loopback.cpp
// @file udp_loopback.cpp
// @brief Portable XPT UDP send/receive over the loopback interface with explicit deadlines.
//
// Hardware-free by design: both endpoints are UDP sockets bound to 127.0.0.1 in this process,
// so the sample runs anywhere XPT builds. Build with WSE_BUILD_XPT=ON and link WSE::Xpt.
//
// XPT deliberately has no default timeout: every operation takes an explicit OperationContext so
// a caller can never wait forever by accident.
//
// This is the one XPT sample whose interesting path always runs, which is why it is also the one
// that returns a non-zero code for a transport failure: with both endpoints inside this process
// there is no absent device or missing server to excuse one. The idle receive at the end is the
// exception, where a Timeout is the result being demonstrated rather than a failure.

#include <xpt/stew.h>
#include <wse/stew.h>

#include <chrono>
#include <cstdint>
#include <string>
#include <vector>

namespace
{

//! Logs a structured transport failure without inventing a category of its own.
void reportFailure( const std::string& operation_in, const wse::xpt::TransportError& error_in )
{
    wse::WLog() << "ERROR:" << operation_in << "failed."
                << "category=" << static_cast< int >( error_in.category() )
                << "code=" << static_cast< int >( error_in.code() )
                << "nativeCode=" << error_in.nativeCode()
                << "message=" << error_in.message();
}

} // namespace

int main()
{
    wse::registDefaultLog();

    // One explicit deadline is reused for every operation in this sample.
    // Reuse is safe because a Timeout carries a duration, not an absolute moment, so each call
    // gets its own second. A second is far more than a loopback datagram needs; it is sized so
    // that a loaded machine cannot make a working exchange look like a failure.
    const wse::xpt::OperationContext context(
        wse::xpt::Timeout( std::chrono::milliseconds( 1000 ) ) );

    // Port zero asks the operating system to assign a free port to each socket.
    // That is what keeps the sample from colliding with anything already running, and it is the
    // one place a zero port is meaningful: as a remote endpoint it would be rejected. The
    // receiver is bound first because a datagram sent to a port nobody holds is simply dropped.
    wse::xpt::UdpClient receiver;
    const wse::xpt::TransportStatus bound_receiver =
        receiver.bind( wse::xpt::Endpoint( "127.0.0.1", 0 ), context );
    if ( !bound_receiver.succeeded() )
    {
        reportFailure( "receiver bind", bound_receiver.error() );
        return 1;
    }

    wse::xpt::UdpClient sender;
    const wse::xpt::TransportStatus bound_sender =
        sender.bind( wse::xpt::Endpoint( "127.0.0.1", 0 ), context );
    if ( !bound_sender.succeeded() )
    {
        reportFailure( "sender bind", bound_sender.error() );
        return 1;
    }

    // The port the operating system chose is only knowable by reading it back; this is the
    // address the sender then aims at. On a closed client the same call returns an invalid
    // endpoint rather than failing.
    const wse::xpt::Endpoint receiver_endpoint = receiver.getLocalEndpoint();
    wse::WLog() << "receiver endpoint:" << receiver_endpoint.host() << ":"
                << receiver_endpoint.port();
    wse::WLog() << "maximum datagram size:" << wse::xpt::UdpClient::maximumDatagramSize();

    // maximumDatagramSize() is the portable ceiling XPT will carry, deliberately below what a
    // particular stack might allow, so a payload that fits is sendable everywhere. A larger one
    // is refused as MessageTooLarge rather than fragmented.
    // A datagram is sent whole or not at all, so the byte count below is either the payload
    // size or nothing; there is no partial send to resume the way a stream has.
    const std::string message = "wse-xpt-udp";
    const std::vector<std::uint8_t> payload( message.begin(), message.end() );
    const wse::xpt::TransferResult<std::size_t> sent =
        sender.sendTo( receiver_endpoint, payload, context );
    if ( !sent.succeeded() )
    {
        reportFailure( "sendTo", sent.error() );
        return 1;
    }
    wse::WLog() << "sent bytes:" << sent.value();

    // Asking for the full ceiling is what makes truncation impossible here. A smaller argument
    // would not wait for a smaller datagram; it would report DatagramTruncated and hand back
    // the leading bytes, with the rest lost, because the remainder of a datagram is discarded
    // rather than queued the way a stream's would be.
    const wse::xpt::TransferResult<wse::xpt::UdpDatagram> received =
        receiver.receiveFrom( wse::xpt::UdpClient::maximumDatagramSize(), context );
    if ( !received.succeeded() )
    {
        reportFailure( "receiveFrom", received.error() );
        return 1;
    }

    // The datagram owns its payload and carries the source endpoint with it, which is the only
    // way an unconnected UDP socket learns who sent what. It is a value held by the result, so
    // it stays valid after the receiver is closed.
    const wse::xpt::UdpDatagram& datagram = received.value();
    const std::string text( datagram.payload().begin(), datagram.payload().end() );
    wse::WLog() << "received from:" << datagram.source().host() << ":"
                << datagram.source().port();
    wse::WLog() << "received payload:" << text;

    // A short deadline with no traffic must report Timeout rather than blocking.
    // Fifty milliseconds is chosen to be plainly shorter than a human notices while still being
    // long enough that a scheduling hiccup cannot make a delivered datagram look late. Nothing
    // is in flight at this point, so the wait is expected to expire; that is the demonstration.
    const wse::xpt::OperationContext short_context(
        wse::xpt::Timeout( std::chrono::milliseconds( 50 ) ) );
    const wse::xpt::TransferResult<wse::xpt::UdpDatagram> idle =
        receiver.receiveFrom( wse::xpt::UdpClient::maximumDatagramSize(), short_context );
    wse::WLog() << "idle receive timed out:"
                << ( !idle.succeeded()
                     && idle.error().code() == wse::xpt::eTransportErrorCode::TimedOut );

    // close() is idempotent; the destructor would do the same. It is not terminal for the
    // object: after it isOpen() is false and getLocalEndpoint() is invalid, but the same client
    // can be bound again.
    receiver.close();
    sender.close();
    // The payload is compared rather than assumed. On loopback within one process a datagram is
    // not realistically lost or reordered, which is what lets this sample assert equality at
    // all; over a real network UDP promises neither, and a protocol would need its own sequence
    // numbers and retries.
    return ( text == message ) ? 0 : 1;
}

File: example/js/xpt/udp_loopback.js
'use strict';

// Portable XPT UDP send and receive over the loopback interface with explicit deadlines.
//
// Hardware-free by design: both endpoints are UDP sockets bound to 127.0.0.1 in this process, so
// the sample runs anywhere XPT builds. Build with WSE_BUILD_XPT=ON and WSE_BUILD_NODE_BINDING=ON.
//
// XPT deliberately has no default timeout: every operation takes an explicit context so a caller
// can never wait forever by accident. These calls block the calling thread until they finish or
// the deadline elapses, so keep the deadline short on the main thread.
//
// Synchronous means synchronous: each call returns a value, not a promise, and nothing else on
// the loop runs while it is in flight. That is why main() below needs no async at all, unlike the
// camera samples. A context may also carry a CancellationSource, but on the main thread there is
// nobody left to call cancel() during the call, so cancellation is only useful in a Worker.
//
// The argument is either the installed `lang/js` package directory or a built `wse.node` addon.
// Given the addon the sample still loads the package, so what is exercised is the public
// JavaScript surface an application uses rather than the raw native entry points.
//
//   node example/js/xpt/udp_loopback.js <installed lang/js directory or wse.node>
//
// This sample takes no arguments of its own.

const path = require('node:path');

const target = process.argv[2] || process.env.WSE_JAVASCRIPT_PACKAGE;
if ( !target ) {
  throw new Error('Pass the installed lang/js directory, or the built wse.node addon.');
}
// A `.node` argument is the build-tree addon. Point the package loader at it and still require the
// package, so the sample always demonstrates the same public surface an application uses.
if ( target.endsWith('.node') ) process.env.WSE_NODE_ADDON = path.resolve(target);
const wse = require(target.endsWith('.node')
  ? path.resolve(__dirname, '..', '..', '..', 'lang', 'js')
  : path.resolve(target));

if ( wse.UdpClient === undefined ) {
  throw new Error('This WSE build does not include the XPT component.');
}

function main() {
  // One explicit deadline is reused for every operation in this sample.
  // Reused, not shared: each call gets a fresh 1000 ms, and the object carries no state between
  // them. A second is far more than a loopback datagram needs and short enough that a mistake
  // does not leave the sample sitting on the main thread.
  const context = { timeoutMs: 1000 };

  // Each object owns a socket from bind() until close(). JavaScript has no deterministic
  // destructor, so nothing releases those sockets on their own: skip close() and they stay bound,
  // holding their ports, until the garbage collector finalizes the objects.
  const receiver = new wse.UdpClient();
  const sender = new wse.UdpClient();
  try {
    // Port zero asks the operating system to assign a free port to each socket.
    // Naming a fixed port instead would make the sample fail wherever that port is already taken,
    // and the assigned port is readable afterwards through localEndpoint().
    // The sender is bound as well, so its datagrams carry a real source the receiver can report.
    receiver.bind('127.0.0.1', 0, context);
    sender.bind('127.0.0.1', 0, context);

    // Only meaningful after the bind: this is where the assigned port becomes known, and it is
    // what the send below aims at.
    const local = receiver.localEndpoint();
    console.log(`receiver endpoint: ${local.host}:${local.port}`);
    // The portable payload ceiling: a datagram larger than this is refused with MessageTooLarge
    // rather than fragmented, because what survives a real network is not what one host allows.
    console.log('maximum datagram size:', wse.UdpClient.maximumDatagramSize());

    const message = Buffer.from('wse-xpt-udp', 'utf8');
    console.log('sent bytes:', sender.sendTo(local.host, local.port, message, context));

    // The ceiling is the maximum rather than the eleven bytes actually expected, because a
    // datagram is delivered whole or not at all: a ceiling below the datagram's size yields
    // DatagramTruncated with a partial payload instead of a smaller successful read. Sizing the
    // buffer by the protocol maximum is the safe default when the sender is not known.
    const datagram = receiver.receiveFrom(wse.UdpClient.maximumDatagramSize(), context);
    // UDP is connectionless, so the source travels with the payload rather than with the socket.
    console.log(`received from: ${datagram.source.host}:${datagram.source.port}`);
    console.log('received payload:', datagram.payload.toString('utf8'));

    sender.sendTo(local.host, local.port, Buffer.from('prefix-and-suffix'), context);
    try {
      receiver.receiveFrom(3, context);
      throw new Error('Expected a truncated datagram');
    } catch (error) {
      if (error.code !== wse.TransportErrorCode.DatagramTruncated) throw error;
      if (error.bytesTransferred !== 3 || error.receivedData.toString() !== 'pre') throw error;
      console.log('truncated prefix:', error.receivedData.toString());
      // sourceEndpoint identifies the sender; the discarded suffix cannot be read again.
    }

    // A short deadline with no traffic must report TimedOut rather than blocking.
    try {
      receiver.receiveFrom(wse.UdpClient.maximumDatagramSize(), { timeoutMs: 50 });
      console.log('idle receive timed out: false');
    } catch (error) {
      console.log('idle receive timed out:', error.code === wse.TransportErrorCode.TimedOut);
    }
  } finally {
    // close() ends the session and is idempotent; the object stays usable for a later open().
    receiver.close();
    sender.close();
  }
}

main();
